llvm/lib/Target/AArch64/AArch64FrameLowering.cpp

```diff
@@ -2497,7 +2497,8 @@ AArch64FrameLowering::getFrameIndexReference(const MachineFunction &MF, int FI,
   return resolveFrameIndexReference(
       MF, FI, FrameReg,
       /*PreferFP=*/
-      MF.getFunction().hasFnAttribute(Attribute::SanitizeHWAddress),
+      MF.getFunction().hasFnAttribute(Attribute::SanitizeHWAddress) ||
+          MF.getFunction().hasFnAttribute(Attribute::SanitizeMemTag),
       /*ForSimm=*/false);
```

Comment:
So this will cause most of the local references to use FP, but our "base tagged pointer" is SP-based (see the aarch64_irg_sp intrinsic). Would not that pessimize the codegen a lot? Please check that nothing stupid is happening in the backend.

There is a bunch of relevant tests under test/CodeGen/AArch64/stack-tagging-*

Reply:
I don't think that makes a difference. I compiled a [moderately complicated file from AOSP](https://cs.android.com/android/platform/superproject/main/+/main:system/unwinding/libunwindstack/Unwinder.cpp) with memtag-stack and without this change (and with and without the `-stack-tagging-record-stack-history=instr`), and the assembly is the same.

The docstring "getFrameIndexReference - Provide a base+offset reference to an FI slot for debug info." also implies this shouldn't affect codegen.

Reply:
Maybe that file was a bad example. Let me check some more.

Reply:
I compiled all of those tests with and without the `getFrameIndexReference` change, the assembly is the same

Without the `getFrameIndexReference` change:
```
for f in llvm/test/CodeGen/AArch64/stack-tagging-*; do build/bin/llc -mtriple=aarch64-linux-android10000 -mattr=+mte -stack-tagging-use-stack-safety=0 -stack-tagging-record-stack-history=instr $f -o tmp2/sp/$(basename $f).S; done
```

and with the `getFrameIndexReference` change:

```
for f in llvm/test/CodeGen/AArch64/stack-tagging-*; do build/bin/llc -mtriple=aarch64-linux-android10000 -mattr=+mte -stack-tagging-use-stack-safety=0 -stack-tagging-record-stack-history=instr $f -o tmp2/fp/$(basename $f).S; done
```

```
for f in fp/*; do diff -u $f sp/$(basename $f); done
```

comes back empty.

I also compiled a semi-randomly picked C++ file from AOSP [Unwinder.cpp](https://cs.android.com/android/platform/superproject/main/+/main:system/unwinding/libunwindstack/Unwinder.cpp?q=Unwinder.cpp&ss=android%2Fplatform%2Fsuperproject%2Fmain) which also compiles to the same Assembly.